### PR TITLE
add support for refactor-open (qualify and unqualify) code action

### PR DESCRIPTION
**Pull request by ulugbekna**
Reservations I have:
- [x] I'm not sure code action names are easily understandable / friendly to newcomers

Blockers: 

- [x] Code actions are shown only if the cursor is on `open` statement, but they can be shown even if the code action won't do anything, e.g.,

```ocaml
open Unix (* both qualify and unqualify code actions will be shown *)
let u = () 
```
- [x] (this is merged into ocaml/merlin#master, we need to rebase vendored merlin) `refactor-open qualify` qualifies with full paths, e.g., for `open Import`, it does `Dune__exe.Import.foo...`.

To-do 
- [x] add tests
- [x] add an entry to the README, so that users know about these code actions

### CODE REVIEW

**Comment by rgrinberg** [comment]
I don't mind the names. This is one of those things that's easy enough to learn by example.

> Code actions are shown only if the cursor is on open statement, but they can be shown even if the code action won't do anything, e.g.,

If the code action won't do anything, the edit it computes should be empty. Can we filter it out then?

**Comment by ulugbekna** [comment]
> If the code action won't do anything, the edit it computes should be empty. Can we filter it out then?

Unfortunately, merlin always returns edits, even if those edits when applied do not change the document at all. 

--- 

Other notes:

1. The code action is quite useful for stdlib and in-file modules refactor-open qualify, as is seen in the tests. However, merlin doesn't use _short paths_ when qualifying modules. Example:

```ocaml
open Import
module Bin = Stdune.Bin

let _PATH =
  lazy
    (Bin.parse_path
       (Dune__exe.Import.Option.value ~default:"" (Dune__exe.Import.Unix_env.get Dune__exe.Import.Unix_env.initial "PATH")))
```

Meanwhile, do you happen to know if that's also the case with refactor-open in merlin's emacs mode?

I will try to fix this anyway.

2. Do you happen to know what's a good way to do several sequenced queries to merlin (vs `ocamlmerlin single`)? I need merlin to take into account ".merlin" and all source files (while `single` query only can take one)?

**Comment by rgrinberg** [comment]
> Meanwhile, do you happen to know if that's also the case with refactor-open in merlin's emacs mode?

No idea. I don't use the emacs mode.

> Do you happen to know what's a good way to do several sequenced queries to merlin (vs ocamlmerlin single)? I need merlin to take into account ".merlin" and all source files (while single query only can take one)?

Is this for tests or something? We don't have .merlin files anymore.

Let's see if @voodoos can help us out.

**Comment by ulugbekna** [comment]
A due update: 

1. I created PRs to the merlin repo for short-paths and useless edits problems (https://github.com/ocaml/merlin/pull/1313 and https://github.com/ocaml/merlin/pull/1314).

I would appreciate your take on whether we should merge this PR as is and wait for upstream merlin to merge fix-PRs, or merge those in ocaml-lsp merlin. Both are fine with me. 

2. 
> > Do you happen to know what's a good way to do several sequenced queries to merlin (vs ocamlmerlin single)? I need merlin to take into account ".merlin" and all source files (while single query only can take one)?
> Is this for tests or something? We don't have .merlin files anymore.

Yes, I needed to write some tests for merlin, but I figured this out.

**Comment by rgrinberg** [comment]
> I would appreciate your take on whether we should merge this PR as is and wait for upstream merlin to merge fix-PRs, or merge those in ocaml-lsp merlin. Both are fine with me.

Do you think this feature is useful to users in its current form?

**Comment by ulugbekna** [comment]
Let's aim to merge this for the release after the next one. I hope by then we'll get a feel whether ocaml/merlin is willing to merge.

**Comment by ulugbekna** [comment]
@rgrinberg I think this PR is ready as soon as we switch to using the latest merlin master. I think that this feature has improved quite a bit in merlin and should be useful to ocaml-lsp users as well :-) 

I included b975615 in this PR as well: 
- what it does? when we update to the latest merlin, we automatically get the fix for #344, so I included the changes entry about it and updated the according test
- why here? we anyway rebase and merge, so including this here should be innocuous; let me know if you'd prefer it in a separate PR 

I can squash all but b975615 for cleaner history; lmk if you'd like that or you have other proposals of course.

File: ocaml-lsp-server/test/e2e/__tests__/textDocument-codeAction.test.ts
Comment (rgrinberg) [inline ocaml-lsp-server/test/e2e/__tests__/textDocument-codeAction.test.ts:545]: This relies on the fact that cmi's of the unix library are available even without the user listing `unix` as an explicit dependency. We don't want our tests to rely on this bug. A true test that uses an external library will tell the server that this library is available.

File: CHANGES.md
Comment (rgrinberg) [inline CHANGES.md:51]: Isn't this already in the merlin change list? We usually do not replicate merlin's change list here.

**Review by rgrinberg — APPROVED** [review]
The PR looks good but we can't rely on the following special libraries in the tests: `unix`, `str`, `dynlink`.

File: CHANGES.md
Comment (ulugbekna) [reply CHANGES.md:51]: It is already in the merlin changes list, but since this bug has been reported several times in both ocaml-lsp & vscode-ocaml-platform, I thought it is useful for users to see it fixed. 

I think that merlin is ocaml-lsp's backend, but the end user shouldn't be lurking in merlin Changes to see what's new in ocaml-lsp -- maybe we could at least add a link to their Changes inside the ocaml-lsp-vendored merlin source.

Comment (ulugbekna) [reply CHANGES.md:51]: The change entry in merlin is "locate: reset global state from all entry points (#1364)", which is cryptic, isn't it?

Comment (rgrinberg) [reply CHANGES.md:51]: It is cryptic. But that can be mended on merlin's side. I'd prefer to see the links to the corresponding change list entries where the fix was done.

> I think that merlin is ocaml-lsp's backend, but the end user shouldn't be lurking in merlin Changes to see what's new in ocaml-lsp -- maybe we could at least add a link to their Changes inside the ocaml-lsp-vendored merlin source.

We could have a policy of replicating only the relevant merlin changes in our repo. The problem is that I often lack context as to what is relevant to lsp, and it is extra work. We can bring it up with the rest of the team to see what they think.

**Review by rgrinberg — APPROVED** [review]

**Pull request merged**
